### DISCUSSION TITLE
[FEAT] Use kitten icat for logo

### DISCUSSION THREAD

**Discussion by hfytr** [fastfetch-cli/fastfetch#1039] [Q&A]
## Wanted features:

I would like to be able to use kitten icat as the fastfetch logo.
## Motivation:

This is because kitten icat works in tmux, while the already implemented kitty direct doesnt work in tmux.

**Comment by CarterLi** [comment]
```
kitten icat --align=left /path/to/logo | fastfetch --raw - --logo-width <width-in-nchars> --logo-height <height-in-nchars>
```

**Reply by hfytr** [reply]
Is there any way to specify this entire thing in the json schema?

**Reply by CarterLi** [reply]
Currently no.

**Reply by barraIhsan** [reply]
So is there a workaround for displaying image inside a tmux (in kitty)?

**Reply by Sk7Str1p3** [reply]
![image](https://github.com/user-attachments/assets/1d93d8c4-65cb-49ec-945a-d4c750eaeb4b)
bruh what the fuck :skull:

**Reply by shadow-absorber** [reply]
> ![image](https://private-user-images.githubusercontent.com/159684143/373970531-1d93d8c4-65cb-49ec-945a-d4c750eaeb4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0MTA0MDUsIm5iZiI6MTczNTQxMDEwNSwicGF0aCI6Ii8xNTk2ODQxNDMvMzczOTcwNTMxLTFkOTNkOGM0LTY1Y2ItNDllYy05NDVhLWQ0Yzc1MGVhZWI0Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyOFQxODIxNDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zY2ExNDcyMTc5MTVlMDk3MTQ3NDllYTEwM2Q1NTExN2JlOTNjMjczZWNiNzIxYjYwYWYyMDZjMWIzN2UzMjMxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Q8XJOGY5cCg8BQpuL1rBtDZgzrCaJgz_5_-meW8ZnA0) bruh what the fuck 💀

you gotta mess with the width and height more... probably lower the height number and increase the width number

**Comment by CarterLi** [comment]
Due to [a kitty bug](https://github.com/kovidgoyal/kitty/issues/7948), you must tell fastfetch how big the image is with this syntax: `--logo-width <width-in-nchars> --logo-height <height-in-nchars>`. `<width-in-nchars>` and `<height-in-nchars>` should be the size that the image actually uses, not the size you want the image to be.

**Comment by shadow-absorber** [comment]
this seems to apply to ghostty too.... so not sure if it is a bug in fastfetch or in ghostty + kitty... considering things like yazi and the kitten icat command works as stated earlier

**Reply by 00-kat** [reply]
> not sure if it is a bug in fastfetch or in ghostty + kitty

Technically, tmux is the one at fault here for refusing to support the Kitty Image Protocol.  It however supports [an escape sequence to pass through sequences to the underlying terminal](https://github.com/tmux/tmux/wiki/FAQ#what-is-the-passthrough-escape-sequence-and-how-do-i-use-it), and the Kitty Image Protocol has a [Unicode placeholders mode](https://sw.kovidgoyal.net/kitty/graphics-protocol/#unicode-placeholders) that, when used in conjunction with that escape sequence, makes images work through tmux (this is what the `icat` kitten does when it detects tmux).  I don't think fastfetch currently supports this.

**Reply by barraIhsan** [reply]
I mean, can't fastfetch use something like [chafa](https://github.com/hpjansson/chafa/) instead? It supports kitten inside tmux

**Reply by CarterLi** [reply]
@00-kat If you can figure out how the tmux escape code works, any information is welcome.

**Reply by shadow-absorber** [reply]
quite sure this is the file that tells kitty:s icat kitten/protocol how to handle tmux placeholders: [https://github.com/kovidgoyal/kitty/blob/master/kittens/icat/main.go](https://github.com/kovidgoyal/kitty/blob/master/kittens/icat/main.go)

**Reply by shadow-absorber** [reply]
line 252 and 253 in [https://github.com/kovidgoyal/kitty/blob/master/kittens/icat/transmit.go](https://github.com/kovidgoyal/kitty/blob/master/kittens/icat/transmit.go) seems to indicate an escape sequence too:
```
	foreground := fmt.Sprintf("\033[38:2:%d:%d:%dm", (imgd.image_id>>16)&255, (imgd.image_id>>8)&255, imgd.image_id&255)
	os.Stdout.WriteString(foreground)
```

**Reply by 00-kat** [reply]
Kitty (and its kittens) are GPL, this project is MIT.  It's probably not permitted to view the source code of the kittens.

I had meant to get around to this but didn't have time, apologies.

**Reply by CarterLi** [reply]
I did try to support tmux, but it didn't work for some reason.

https://github.com/fastfetch-cli/fastfetch/blob/a378d2ad6234e23dad903bea6e287c005a376e09/src/logo/image/image.c#L189-L199

PR is welcome if you can fix it.

**Reply by 00-kat** [reply]
I wasn't able to get the tmux pass-through escape sequence to work either :/.

**Comment by CarterLi** [comment]
Support for rendering with `kitten icat` is added in d58d786c83141f030df00faf5d8478a292455fb7

`fastfetch --kitty-icat ~/earth.gif`

![image](https://github.com/user-attachments/assets/09c0f3f2-89e7-466b-a16c-f6ccc6af683c)

Tested with `kitty` + `tmux` + [`gif`](https://commons.wikimedia.org/wiki/File:Rotating_earth_(large).gif)

**Reply by CarterLi** [reply]
Requires `kitten` is available in `$PATH`, of course.

**Reply by shadow-absorber** [reply]
tested using the aur fasfetch-git package and works like a charm

**Reply by jasonshanks** [reply]
Is this fix in the stable release yet? (using home-brew package).

I discovered this was the issue for me too (in Ghostty terminal), tried: kitty, kitty-direct, kitty-icat

Also tried adding this to tmux configs:
`set-option -g allow-passthrough on `

**Reply by CarterLi** [reply]
> Is this fix in the stable release yet? (using home-brew package).

`kitty-icat` isn't released yet. No change to others.

**Reply by barraIhsan** [reply]
Confirm working using the `fastfetch-git` aur package, do note that the image will be cropped off if the height of the image is greater than the text height on tmux
`"type": "kitty"` (only work outside tmux)  
<details><summary>Image</summary>

![image](https://github.com/user-attachments/assets/87d9e65c-e4f8-456f-885a-3b2953222a40)

</details>

`"type": "kitty-icat"` (outside tmux)
<details><summary>Image</summary>

![image](https://github.com/user-attachments/assets/4a913b15-fc6c-468f-b818-44a34555292e)

</details>

`"type": "kitty-icat"` (inside tmux)
<details><summary>Image</summary>

![image](https://github.com/user-attachments/assets/738b7046-6aad-486b-8976-e5f23a2df7e2)

</details>

**Comment by linkarzu** [comment]
This was working wonderfully in TMUX, but a few months ago, it stopped working, I asked other users in my discord and they also have the same issue, anybody else?
My config can be found here 
https://github.com/linkarzu/dotfiles-latest/blob/main/fastfetch/config.jsonc

I'm using `kitty-icat` when inside tmux
<div align="center">
    <img src="https://github.com/user-attachments/assets/8f482ceb-3c76-4dd6-84d7-a51a3881fafe" width="700">
</div>

Here I'm outside tmux just using `kitty` and it works
<div align="center">
    <img src="https://github.com/user-attachments/assets/e017aee7-c08b-4262-80ab-c1d9ab4e7ad5" width="700">
</div>

**Reply by mackrus** [reply]
Saw your video, thought you had the solution, checked your dotfiles, you had the same setup :( the circle is now complete. 
kitty-icat seems to work inside kitty though, but not in ghostty

**Reply by linkarzu** [reply]
Quick question for you. Other users in discord says it works for them in tmux, buuuut, they use M1, or M3 macbooks. 
I have an M4 macbook, and there's another user that also has an M4 having the same issue. What hardware do you use?

**Reply by linkarzu** [reply]
Just for context, here's the video:
https://youtu.be/9OoO2t6TOY8

**Reply by mackrus** [reply]
i'm using an intel-based macbook

**Reply by shadow-absorber** [reply]
on an amd proccessor on arch linux here and it still works... might be because shadow has both ghostty and kitty installed

**Reply by linkarzu** [reply]
So it seems macOS related, I cannot confirm about intel macs, but 1 other person has confirmed for me that it does not work for them using an M4, still not sure yet if only M4 related or what. M1s and M3s don't seem to have issues

**Reply by JoshSald** [reply]
On my M4 can confirm that this issue persists. @linkarzu have you managed to find a solution?

**Reply by linkarzu** [reply]
@JoshSald no solution, thanks for confirming you have an issue with M4 as well, I created an issue:
#1866
